cloudmarker/test/test_esstore.py

```diff
@@ -0,0 +1,73 @@
+"""Tests for EsStore plugin."""
+
+import unittest
+from unittest import mock
+
+from elasticsearch import ElasticsearchException
+
+from cloudmarker.stores import esstore
+
+
+class EsStoreTest(unittest.TestCase):
+    """Tests for EsStore plugin."""
+
+    @mock.patch('cloudmarker.stores.esstore.Elasticsearch.bulk')
+    def test_happy_flow(self, mock_bulk):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        es_store = esstore.EsStore()
+
+        mock_bulk_response = {
+            'errors': '',
+            'items': [
+                {'index': {'status': 200}}
+            ]
+        }
+
+        mock_bulk.return_value = mock_bulk_response
+
+        es_store.write(mock_record)
+        es_store.done()
+
+        mock_bulk.assert_called_once()
+        # In case of successful push the buffer should reset.
+        self.assertEqual(es_store._buffer, '')
+
+    @mock.patch('cloudmarker.stores.esstore.Elasticsearch.bulk')
+    def test_exception_flow(self, mock_bulk):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        es_store = esstore.EsStore()
+
+        mock_bulk.side_effect = ElasticsearchException()
+
+        es_store.write(mock_record)
+        es_store.done()
+
+        # In case of any expception the buffer should not reset.
+        self.assertNotEqual(es_store._buffer, '')
+
+    @mock.patch('cloudmarker.stores.esstore.Elasticsearch.bulk')
+    def test_bulk_response_error(self, mock_bulk):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        es_store = esstore.EsStore()
+
+        mock_bulk_response = {
+            'errors': 'foo',
+            'items': [
+                {
+                    'index': {
+                        'status': 500,
+                        '_id': 'foo'
+                    }
+                }
+            ]
+        }
+
+        mock_bulk.return_value = mock_bulk_response
+
+        es_store.write(mock_record)
+        es_store.done()
+
+        mock_bulk.assert_called_once()
```

Comment:
I'd mock the logger here as well.

```
@mock.patch('cloudmarker.stores.esstore._log')
    def test_bulk_response_error(self, mock_logger, mock_bulk):
```

```
mock_logger.debug.assert_called_with('Failed to insert record; ID: %s', 'foo')
mock_logger.error.assert_called_with('Failed to write %d records', 1)
```

Reply:
I suggest not testing info and debug logs. The logs are meant for diagnostics which is a bonus addition to the documented behaviour of the project. I prefer not testing these bonus additions. Having said that, I think there is value in testing warning and error logs for exceptional conditions (such as an exception). Logging warning or error for exceptional conditions is a necessary feature and we want to be sure that these logs accidentally do not disappear in future commits, so having tests for these sounds like a good idea.

Reply:
Can we find some other ways to test to identify if the exceptions are raised or not? Checking the values for the loggers is not very much convincing to me. I also initially thought the same.

Reply:
If we have `mock_bulk.side_effect = ElasticsearchException()` in the test code, then the exception is definitely being raised. There is no need to test if the exception is raised. That would amount to only testing whether `unittest.mock` is working as expected. We don't want to test `unittest.mock`. We want to test `EsStore`.

There is no need to explicitly test the exception here because the raising of exception and handling it is internal behavior in the current `EsStore`. The external behavior is just that `EsStore` does not crash which the existing tests are testing. There is of course logging as another external behaviour but I guess we are deciding here not to test that.

Finally, how can we be sure that our test itself is correct and it did trigger an exception which was handled by the `except` block? For that, we have code coverage. For example, see this [code coverage report for esstore.py](https://coveralls.io/builds/22962027/source?filename=cloudmarker/stores/esstore.py) triggered due to this pull request. It shows that lines 70-74, i.e., the `except` block were exercised by our test. So the code coverage report proves that our tests were able to trigger an exception that `EsStore` could handle.